source/openvr/openvr.cpp

```diff
@@ -29,13 +32,112 @@ static inline vr::VRTextureBounds_t calc_side_by_side_bounds(vr::EVREye eye, con
 	vr::VRTextureBounds_t bounds = (eye != vr::Eye_Right) ?
 		vr::VRTextureBounds_t { 0.0f, 0.0f, 0.5f, 1.0f } : // Left half of the texture
 		vr::VRTextureBounds_t { 0.5f, 0.0f, 1.0f, 1.0f };  // Right half of the texture
+
 	if (orig_bounds != nullptr && orig_bounds->uMin > orig_bounds->uMax)
 		std::swap(bounds.uMin, bounds.uMax);
 	if (orig_bounds != nullptr && orig_bounds->vMin > orig_bounds->vMax)
 		std::swap(bounds.vMin, bounds.vMax);
+
 	return bounds;
 }
 
+/*
+Get the device api based on the texture type
+*/
+static inline boolean isD3D10(void *pTexture)
```

Comment:
I actually had a simpler approach before similar to yours but even DX11 textures could be cast to dx DX10 without failure. I could only go through the reshade device impl stored in the private data to really find if it is the right render api. Because of the complexity of the the check it will use the swapbuffer when it is resolved to speed things up.

Reply:
Mmh. I think these days the D3D10 driver is really just a wrapper around D3D11, so it may be that querying a D3D10 texture can also give you the D3D11 one.
In that case this should still be possible to do, just need to first query for `ID3D10Texture2D` and then for `ID3D11Texture2D` (I don't think the D3D11 driver gives you the D3D10 object), rather than in the order shown above (did it in this order, since there are vastly more D3D11 games out there, so would have been nice to avoid the cost of querying for D3D10 for those).

Reply:
Sadly it does. I tried it first to cast to the 'ID3D10Texture2D' and even when it was a DX11 one it passed through. The previous code crashed on DX10 accessing the proxy fetched from the device. All other stuff went before went through in the dx11 submit call.

Reply:
I see. I'm a bit confused, since the OpenVR headers state that `TextureType_DirectX` should be `ID3D11Texture2D` (https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h#L158). So not sure how that can be `ID3D10Texture2D`. But maybe that is just documented wrong.

It does look like the current code should crash in D3D11 though. When a `ID3D11Texture2D` object is passed into `isD3D10` it's statically cast to `ID3D10Texture2D`, even though it is not that. This by sheer luck happens to work because both `ID3D10Texture2D` and `ID3D11Texture2D` have the `GetDevice` method at the same vtable index (3) and since the vtable pointer is stored at the beginning it will make use of the `ID3D11Texture2D` method regardless of the cast (the compiler translates ```dxTexture->GetDevice(&device);``` to ```void **vtable = *(void***)dxTexture; ((void(*)(void *, void **))vtable[3])(dxTexture, &device);```). But now `device` actually points to a `ID3D11Device` object and not a `ID3D10Device` object as the code assumes. This means the `GetPrivateData` call should crash, since it is located at a different vtable index in `ID3D10Device` (76) vs `ID3D11Device` (34). There is no entry for index 76 in the `ID3D11Device` vtable, so that will call an invalid address. Only way to ensure you get the correct object would be to first `QueryInterface` on the device (since `QueryInterface` is always at vtable index 0 and therefore safe to call).

But I think this can be abused correctly like this (assuming `QueryInterface` on a D3D11 device does not return a D3D10 device, and from a quick test of mine this does not appear to be the case):
```c++
static vr::EVRCompositorError on_submit_d3d10(vr::EVREye eye, ID3D10Texture2D *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags,
	std::function<vr::EVRCompositorError(vr::EVREye eye, void *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags)> submit)
{
	com_ptr<ID3D10Device> device;
	texture->GetDevice(&device);

	D3D10Device *device_proxy = nullptr; // Was set via 'SetPrivateData', so do not use a 'com_ptr' here, since 'GetPrivateData' will not add a reference
	if (UINT data_size = sizeof(device_proxy);
		FAILED(device->GetPrivateData(__uuidof(D3D10Device), &data_size, reinterpret_cast<void *>(&device_proxy))))
		goto normal_submit; // No proxy device found, so just submit normally
	...
}
static vr::EVRCompositorError on_submit_d3d11(vr::EVREye eye, ID3D11Texture2D *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags,
	std::function<vr::EVRCompositorError(vr::EVREye eye, void *texture, const vr::VRTextureBounds_t *bounds, vr::EVRSubmitFlags flags)> submit)
{
	com_ptr<ID3D11Device> device;
	texture->GetDevice(&device); // 'GetDevice' is at the same vtable index for 'ID3D10Texture2D' and 'ID3D11Texture2D', so this happens to work for either case

	if (com_ptr<ID3D10Device> device10;
		device->QueryInterface(&device10) == S_OK)
		// Whoops, this is actually a D3D10 texture, redirect ...
		return on_submit_d3d10(eye, reinterpret_cast<ID3D10Texture2D *>(texture), bounds, flags);

	D3D11Device *device_proxy = nullptr; // Was set via 'SetPrivateData', so do not use a 'com_ptr' here, since 'GetPrivateData' will not add a reference
	if (UINT data_size = sizeof(device_proxy);
		FAILED(device->GetPrivateData(__uuidof(D3D11Device), &data_size, reinterpret_cast<void *>(&device_proxy))))
		goto normal_submit; // No proxy device found, so just submit normally
	...
}

...

case vr::TextureType_DirectX:
	return on_submit_d3d11(eEye, static_cast<ID3D11Texture2D *>(pTexture->handle), pBounds, nSubmitFlags, submit_lambda);
```

Reply:
No you are right that openVR has dokumented that it is expected to get a DX11 texture. But they also have DX10 support and even worked on DX9 but that implementation never got finished. Raceroom might be the only game using DX10 for VR. Was surprised that it was possible/supported as well.

If your example code works it would be more slick and cleaner solution. But I think i tried to use QueryInterface() and it did not work as expected. But I might not did it correctly. I will try it again and when it works rework the code.

Reply:
Would be interesting to know if it's actually a `ID3D10Texture2D` object passed in. You can check this in the Visual Studio debugger by putting a breakpoint in the submit function and when that breaks, entering `((IUnknown*)texture)->__vfptr` in the Watch window. The Value column will then show you the name of the vtable (when symbols are loaded).

If this is actually a `ID3D11Texture2D`, then this is
```
{d3d11.dll!const CLayeredObjectWithCLS<class CTexture2D>::CContainedObject::`vftable'{for `CDeviceChild<struct ID3D11Texture2D>'}} {...}
```
If this is actually a `ID3D10Texture2D`, then this is something else, e.g.
```
{d3d11.dll!const CLayeredObjectWithCLS<class CTexture2D>::CContainedObject::`vftable'} {...}
```

Reply:
Damn Raceroom seems very twitchy and crash quickly when attaching the debugger... On beatsaber debugging was no issue at all. Before it crashes it also tells that the brakepoints does not match the code. Are there some tricks for this or a way to output the ((IUnknown*)texture)->__vfptr into the log file?

Reply:
Great it works by fetching the device interface as you suggested. Simplified this now. Thx.